cpp/CMakeLists.txt

```diff
@@ -783,24 +543,133 @@ set_property(CACHE RMM_LOGGING_LEVEL PROPERTY STRINGS
+    add_dependencies("${NAMESPACE}_${MODULE}" stringify_run)
+    target_compile_definitions("${NAMESPACE}_${MODULE}" PUBLIC "SPDLOG_ACTIVE_LEVEL=SPDLOG_LEVEL_${LOGGING_LEVEL}")
+    add_library("${NAMESPACE}::${MODULE}" ALIAS "${NAMESPACE}_${MODULE}")
+endfunction()
+
+set(CUDF_NAMESPACE "cudf")
+set(CUDF_SRC "${CMAKE_BINARY_DIR}/cudf.cu")  # this file will be generated
+
+set(
+    CUDF_HEADERS_DIRS
+        "${CMAKE_SOURCE_DIR}/include/cudf"
+        "${CMAKE_SOURCE_DIR}/include/nvtext"
+)
+
+set(
+    CUDF_BASE_MODULE_LIST
+        "${CMAKE_SOURCE_DIR}/src/aggregation"
+        "${CMAKE_SOURCE_DIR}/src/binaryop"
+        "${CMAKE_SOURCE_DIR}/src/bitmask"
+        "${CMAKE_SOURCE_DIR}/src/column"
+        "${CMAKE_SOURCE_DIR}/src/copying"
+        "${CMAKE_SOURCE_DIR}/src/datetime"
+        "${CMAKE_SOURCE_DIR}/src/dictionary"
+        "${CMAKE_SOURCE_DIR}/src/filling"
+        "${CMAKE_SOURCE_DIR}/src/groupby"
+        "${CMAKE_SOURCE_DIR}/src/jit"
+        "${CMAKE_SOURCE_DIR}/src/lists"
+        "${CMAKE_SOURCE_DIR}/src/quantiles"
+        "${CMAKE_SOURCE_DIR}/src/reshape"
+        "${CMAKE_SOURCE_DIR}/src/scalar"
+        "${CMAKE_SOURCE_DIR}/src/search"
+        "${CMAKE_SOURCE_DIR}/src/sort"
+        "${CMAKE_SOURCE_DIR}/src/stream_compaction"
+        "${CMAKE_SOURCE_DIR}/src/strings"
+        "${CMAKE_SOURCE_DIR}/src/structs"
+        "${CMAKE_SOURCE_DIR}/src/table"
+        "${CMAKE_SOURCE_DIR}/src/text"
+        "${CMAKE_SOURCE_DIR}/src/transform"
+        "${CMAKE_SOURCE_DIR}/src/unary"
+        "${CMAKE_SOURCE_DIR}/src/utilities"
+)
+
+set(
+    CUDF_MODULE_LIST
+        "${CMAKE_SOURCE_DIR}/src/ast"
+        "${CMAKE_SOURCE_DIR}/src/comms"
+        "${CMAKE_SOURCE_DIR}/src/hash"
+        "${CMAKE_SOURCE_DIR}/src/interop"
+        "${CMAKE_SOURCE_DIR}/src/io"
+        "${CMAKE_SOURCE_DIR}/src/join"
+        "${CMAKE_SOURCE_DIR}/src/merge"
+        "${CMAKE_SOURCE_DIR}/src/partitioning"
+        "${CMAKE_SOURCE_DIR}/src/reductions"
+        "${CMAKE_SOURCE_DIR}/src/replace"
+        "${CMAKE_SOURCE_DIR}/src/reshape"
+        "${CMAKE_SOURCE_DIR}/src/rolling"
+        "${CMAKE_SOURCE_DIR}/src/transpose"
+)
+
+foreach(CUDF_BASE_MODULE_DIR IN LISTS CUDF_BASE_MODULE_LIST)
+    file(GLOB_RECURSE CUDF_BASE_FILES "${CUDF_BASE_MODULE_DIR}/*")
+    list(APPEND CUDF_BASE_SRC ${CUDF_BASE_FILES})
+endforeach()
+
+add_library("${CUDF_NAMESPACE}_base" ${CUDF_BASE_SRC})
+add_library("${CUDF_NAMESPACE}::base" ALIAS "${CUDF_NAMESPACE}_base")
+add_dependencies("${CUDF_NAMESPACE}_base" stringify_run)
+
+foreach(CUDF_MODULE_DIR IN LISTS CUDF_MODULE_LIST)
+    string(  # replace "/" with ";" to create a list of directory leaves for the module directory
+        REPLACE "/" ";"
+            MODULE_DIR_LEAVES
+            ${CUDF_MODULE_DIR}
+    )
+    list(REVERSE MODULE_DIR_LEAVES)
+    list(GET MODULE_DIR_LEAVES 0 CUDF_MODULE_NAME)  # name the module after the last directory leaf
```

Comment:
It was hard to wrap my brain around this code. I think we should either try to use existing CMake functions like `get_filename_component` or abstract this into a function with a comment describing what it does.

Reply:
I played with `get_filename_component`, but to get the directories, you have to play a lot with relative paths. That felt confusing to me, personally.

If I were to comment this code more, it would be:
1. Use replacement regex to transform `/` with `;` forcing CMake to reinterpret the variable as a list
2. Reverse the order of the list so that the directory under `cpp/src` is first
3. Grab that directory name and call it the module name

Reply:
I think a `get_filename_component` pattern would be helpful to understand, but adding the 3 points you have above as comments as well as the big picture goal of the function would be helpful.